File: proservicios/src/Servicios/PagoService.php
<?php
require_once __DIR__ . '/../../config/Database.php';

class PagoService {
    private $conn;

    public function __construct() {
        $database = new Database();
        $this->conn = $database->getConnection();
    }

    // 1. OBTENER DATOS PARA PAGAR (Pantalla de Checkout)
    public function obtenerReservaParaPago($reserva_id, $usuario_id) {
        $sql = "SELECT r.*, s.nombre_servicio, s.descripcion, u.nombre, u.apellido, u.email, u.telefono 
                FROM reservas r
                JOIN servicios s ON r.servicio_id = s.servicio_id
                JOIN usuarios u ON r.usuario_id = u.usuario_id
                WHERE r.reserva_id = :rid AND r.usuario_id = :uid";
        
        $stmt = $this->conn->prepare($sql);
        $stmt->bindParam(':rid', $reserva_id);
        $stmt->bindParam(':uid', $usuario_id);
        $stmt->execute();
        return $stmt->fetch(PDO::FETCH_ASSOC);
    }

    // 2. REGISTRAR EL PAGO
    public function registrarPago($reserva_id, $monto, $metodo) {
        try {
            $this->conn->beginTransaction();

            $sqlPago = "INSERT INTO pagos (reserva_id, monto, metodo_pago, estado_pago) VALUES (:rid, :monto, :metodo, 'aprobado')";
            $stmt = $this->conn->prepare($sqlPago);
            $stmt->bindParam(':rid', $reserva_id);
            $stmt->bindParam(':monto', $monto);
            $stmt->bindParam(':metodo', $metodo);
            $stmt->execute();

            $sqlReserva = "UPDATE reservas SET estado = 'pagada' WHERE reserva_id = :rid";
            $stmtRes = $this->conn->prepare($sqlReserva);
            $stmtRes->bindParam(':rid', $reserva_id);
            $stmtRes->execute();

            $this->conn->commit();
            return ['success' => true, 'message' => 'Pago registrado'];

        } catch (Exception $e) {
            $this->conn->rollBack();
            return ['success' => false, 'message' => $e->getMessage()];
        }
    }

    // 3. HISTORIAL DE PAGOS (Lista)
    public function obtenerHistorialPorUsuario($usuario_id) {
        $sql = "SELECT p.*, s.nombre_servicio, r.fecha_reserva, r.reserva_id
                FROM pagos p
                JOIN reservas r ON p.reserva_id = r.reserva_id
                JOIN servicios s ON r.servicio_id = s.servicio_id
                WHERE r.usuario_id = :uid
                ORDER BY p.pago_id DESC";
        
        $stmt = $this->conn->prepare($sql);
        $stmt->bindParam(':uid', $usuario_id);
        $stmt->execute();
        return $stmt;
    }

    // 4. OBTENER COMPROBANTE INDIVIDUAL 
    public function obtenerComprobante($reserva_id) {
        // Traemos datos del pago, reserva, servicio, cliente y proveedor
        $sql = "SELECT 
                    p.pago_id, p.monto, p.metodo_pago, p.estado_pago,
                    r.fecha_reserva, r.horario_elegido, r.reserva_id,
                    s.nombre_servicio, s.descripcion, s.precio,
                    u.nombre as cliente_nombre, u.apellido as cliente_apellido, u.email as cliente_email,
                    prov.nombre as prov_nombre, prov.apellido as prov_apellido, prov.email as prov_email
                FROM pagos p
                JOIN reservas r ON p.reserva_id = r.reserva_id
                JOIN servicios s ON r.servicio_id = s.servicio_id
                JOIN usuarios u ON r.usuario_id = u.usuario_id
                JOIN usuarios prov ON s.proveedor_id = prov.usuario_id
                WHERE r.reserva_id = :rid";

        $stmt = $this->conn->prepare($sql);
        $stmt->bindParam(':rid', $reserva_id);
        $stmt->execute();
        return $stmt->fetch(PDO::FETCH_ASSOC);
    }
}
?>

File: qa/src/pago_service.py

class PagoService:
    def __init__(self, db_mock=None, notificador_mock=None):
        self.db = db_mock or self._crear_db_mock()
        self.notificador = notificador_mock or self._crear_notificador_mock()

    def _crear_db_mock(self):
        return {
            "reservas": {
                101: {"reserva_id": 101, "usuario_id": 501, "servicio_id": 1, "estado": "pendiente"},
                102: {"reserva_id": 102, "usuario_id": 502, "servicio_id": 2, "estado": "pendiente"},
            },
            "pagos": [],
            "usuarios": {
                501: {"nombre": "Ana", "email": "ana@example.com"},
                502: {"nombre": "Luis", "email": "luis@example.com"},
            },
            "servicios": {
                1: {"nombre_servicio": "Masaje", "precio": 50.0},
                2: {"nombre_servicio": "Corte", "precio": 20.0},
            }
        }

    def _crear_notificador_mock(self):
        return NotificadorMock()

    def registrar_pago(self, reserva_id, monto, metodo):
        try:
            if reserva_id not in self.db["reservas"]:
                raise ValueError("Reserva no encontrada")

            reserva = self.db["reservas"][reserva_id]
            usuario_id = reserva["usuario_id"]

            nuevo_pago = {
                "pago_id": len(self.db["pagos"]) + 1,
                "reserva_id": reserva_id,
                "monto": monto,
                "metodo": metodo,
                "estado": "aprobado"
            }
            self.db["pagos"].append(nuevo_pago)

            self.db["reservas"][reserva_id]["estado"] = "pagada"

            self._notificar_usuario(reserva_id, monto, usuario_id)

            return {"success": True}

        except Exception as e:
            return {"success": False, "message": str(e)}

    def _notificar_usuario(self, reserva_id, monto, usuario_id):
        if usuario_id in self.db["usuarios"]:
            usuario = self.db["usuarios"][usuario_id]
            mensaje = f"Hola {usuario['nombre']}, tu pago de ${monto} para la reserva #{reserva_id} ha sido confirmado."
            self.notificador.enviar_email(usuario["email"], "Pago Exitoso", mensaje)

    def obtener_comprobante(self, reserva_id, usuario_id):
        if reserva_id not in self.db["reservas"]:
            return None
        reserva = self.db["reservas"][reserva_id]
        if reserva["usuario_id"] != usuario_id:
            return None

        servicio = self.db["servicios"][reserva["servicio_id"]]
        usuario = self.db["usuarios"][usuario_id]
        pago = next((p for p in self.db["pagos"] if p["reserva_id"] == reserva_id), None)

        if not pago:
            return None

        return {
            "pago_id": pago["pago_id"],
            "monto": pago["monto"],
            "metodo_pago": pago["metodo"],
            "fecha_reserva": "2025-12-14",
            "reserva_id": reserva_id,
            "nombre_servicio": servicio["nombre_servicio"],
            "nombre": usuario["nombre"],
            "apellido": "",
            "email": usuario["email"]
        }


class NotificadorMock:
    def __init__(self):
        self.enviados = []

    def enviar_email(self, email, asunto, mensaje):
        self.enviados.append({"email": email, "asunto": asunto, "mensaje": mensaje})

#actualizando prueba
